### PR TITLE
[message-buffer 10/X] - Message Buffer Admin IXs & variable length

### PR DESCRIPTION
**Pull request by swimricky**
### Summary
Add `create/resize/delete_buffer` ixs, allow for `MessageBuffer` to be variable length

### Changes
- Remove `fund`
- Rename `whitelist.authority` to `whitelist.admin` and check that admin is signer & payer for all admin ixs.
- `MessageBuffer` struct now only defines the header fields and allows for variable length for the actual messages (see `MessageBuffer::put_all_in_buffer`)

### CODE REVIEW

File: message_buffer/programs/message_buffer/src/instructions/resize_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    },
+                ),
+                target_rent - message_buffer_account_info.lamports(),
+            )?;
+        }
+        message_buffer_account_info
+            .realloc(target_size, false)
+            .map_err(|_| MessageBufferError::ReallocFailed)?;
+    } else {
+        // TODO:
```
Comment (swimricky) [inline message_buffer/programs/message_buffer/src/instructions/resize_buffer.rs:76]: @jayantk what are your thoughts on this?

File: message_buffer/programs/message_buffer/src/lib.rs [hunk 2]
```diff
@@ -84,6 +82,47 @@ pub mod message_buffer {
...
+        instructions::resize_buffer(
+            ctx,
+            allowed_program_auth,
+            base_account_key,
+            buffer_bump,
+            target_size,
+        )
+    }
+
+    // TODO: should delete_buffer have same CPI restrictions
```
Comment (swimricky) [inline message_buffer/programs/message_buffer/src/lib.rs:116]: @jayantk do we want to be able to call `delete_buffer` outside of a pyth `delete_XXX` ix?

File: message_buffer/programs/message_buffer/src/instructions/create_buffer.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            Transfer,
+        },
+    },
+};
+
+pub fn create_buffer<'info>(
+    ctx: Context<'_, '_, '_, 'info, CreateBuffer<'info>>,
+    allowed_program_auth: Pubkey,
+    base_account_key: Pubkey,
+    target_size: u32,
```
Comment (swimricky) [inline message_buffer/programs/message_buffer/src/instructions/create_buffer.rs:23]: @jayantk for `target_size` do you think we should implement it in this way or should we assume the hard coded max of 10KB?

File: message_buffer/programs/message_buffer/src/lib.rs [hunk 4]
```diff
@@ -84,6 +82,47 @@ pub mod message_buffer {
     ) -> Result<()> {
         instructions::put_all(ctx, base_account_key, messages)
     }
+
+
+    /// Initializes the buffer account with the target_size
+    ///
+    /// TODO: should target_size be a parameter or should
```
Comment (swimricky) [inline message_buffer/programs/message_buffer/src/lib.rs:89]: @jayantk  what are your thoughts on this?

File: message_buffer/programs/message_buffer/src/instructions/resize_buffer.rs (hunk 1 above)
Comment (jayantk) [reply message_buffer/programs/message_buffer/src/instructions/resize_buffer.rs:76]: I think it's fine to shrink and not worry about the lamports. just leave a comment that the account will retain more rent than needed

File: message_buffer/programs/message_buffer/src/lib.rs (hunk 4 above)
Comment (jayantk) [reply message_buffer/programs/message_buffer/src/lib.rs:86]: target_size parameter is good. I think we have to manage the sizes anyway

File: message_buffer/programs/message_buffer/src/lib.rs (hunk 2 above)
Comment (jayantk) [reply message_buffer/programs/message_buffer/src/lib.rs:113]: yeah let the admin delete whatever.

can you add doc comments to all of these instructions?

File: message_buffer/programs/message_buffer/src/state/message_buffer.rs [hunk 5]
```diff
@@ -1,36 +1,31 @@
...
 
-
-/// `MessageBuffer` is an arbitrary set of bytes
-/// that will be included in the AccumulatorSysvar
-///
+/// A MessageBuffer will have the following structure
+/// ```ignore
+/// struct MessageBuffer {
+///     header: BufferHeader,
+///     messages: [u8; accountInfo.data.len - header.header_len]
```
[the remark this answers is not in the document]
Comment (jayantk) [reply message_buffer/programs/message_buffer/src/state/message_buffer.rs:13]: this whole comment is confusing. I think you want to say that the MessageBuffer struct *is* effectively the header, and then is followed by more space in the account for the messages themselves.

File: message_buffer/programs/message_buffer/src/state/message_buffer.rs [hunk 6]
```diff
@@ -247,40 +277,52 @@ mod test {
         ];
 
         let data_bytes: Vec<Vec<u8>> = data.into_iter().map(data_bytes).collect();
+        // let message_buffer = &mut MessageBufferTemp::new(0);
+        // let (num_msgs, num_bytes) = message_buffer.put_all(&data_bytes);
```
Comment (jayantk) [inline message_buffer/programs/message_buffer/src/state/message_buffer.rs:281]: delete this

File: message_buffer/tests/message_buffer.ts [hunk 7]
```diff
@@ -1,28 +1,38 @@
...
 const [mockCpiCallerAuth] = anchor.web3.PublicKey.findProgramAddressSync(
   [messageBufferProgram.programId.toBuffer(), Buffer.from("cpi")],
   mockCpiProg.programId
 );
-const [fundPda] = anchor.web3.PublicKey.findProgramAddressSync(
-  [Buffer.from("fund")],
-  messageBufferProgram.programId
-);
+
+console.log("2");
```
Comment (jayantk) [inline message_buffer/tests/message_buffer.ts:35]: delete these extraneous log statements (there are a few of them)

File: message_buffer/tests/message_buffer.ts [hunk 8]
```diff
@@ -100,29 +151,132 @@ describe("accumulator_updater", () => {
...
+    );
+    const messageBufferHeader = deserializeMessageBufferHeader(
+      messageBufferProgram,
+      messageBufferAccountData
+    );
+    assert.equal(messageBufferHeader.version, 1);
+    assert.equal(messageBufferHeader.bump, accumulatorPdaBump);
+  });
+
+  it("Creates a buffer even if the account already has lamports", async () => {
```
Comment (jayantk) [inline message_buffer/tests/message_buffer.ts:186]: good test

File: message_buffer/tests/message_buffer.ts [hunk 9]
```diff
@@ -100,29 +151,132 @@ describe("accumulator_updater", () => {
...
+    );
+    assert.isTrue(whitelist.admin.equals(whitelistAdmin.publicKey));
   });
 
   it("Mock CPI program - AddPrice", async () => {
     const mockCpiCallerAddPriceTxPubkeys = await mockCpiProg.methods
       .addPrice(addPriceParams)
       .accounts({
-        fund: fundPda,
+        // fund: fundPda,
```
Comment (jayantk) [inline message_buffer/tests/message_buffer.ts:279]: delete

File: message_buffer/tests/message_buffer.ts [hunk 10]
```diff
@@ -413,23 +751,39 @@ export const getAccumulatorPdaMeta = (
...
 
-type BufferHeader = IdlTypes<MessageBuffer>["BufferHeader"];
+async function getMessageBuffer(
+  connection: anchor.web3.Connection,
+  accountKey: anchor.web3.PublicKey
+): Promise<Buffer | null> {
+  let accountInfo = await connection.getAccountInfo(accountKey);
+  return accountInfo ? accountInfo.data : null;
+  // return (accountInfo ?? accountInfo.data)
+  // return (await connection.getAccountInfo(accumulatorPdaKey)).data;
```
Comment (jayantk) [inline message_buffer/tests/message_buffer.ts:761]: delete all the commented out code

**Review by jayantk — APPROVED** [review]
Generally looks good to me with a bunch of minor comments. I think in the interest of time, please fix the API (specifically the unnecessary system_program args) so we can integrate, merge, and then address feedback in a subsequent PR.

File: message_buffer/programs/message_buffer/src/instructions/delete_buffer.rs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    allowed_program_auth: Pubkey,
+    base_account_key: Pubkey,
+    bump: u8,
+) -> Result<()> {
+    let message_buffer_account_info = ctx
+        .remaining_accounts
+        .first()
+        .ok_or(MessageBufferError::MessageBufferNotProvided)?;
+
+    verify_message_buffer(message_buffer_account_info)?;
```
[the remark this answers is not in the document]
Comment (swimricky) [reply message_buffer/programs/message_buffer/src/instructions/delete_buffer.rs:26]: close won't be an issuue. The name is a bit misleading. I'm renaming this to `MessageBuffer::check_discriminator` for clarity which shouldn't fail except for actual invalid cases.

**Pull request merged**